tests/retrieval/helpers.py

```diff
@@ -1,129 +1,391 @@
+    empty_target_action: str = "skip",
+    **kwargs
 ) -> Tensor:
     """ Compute metric with multiple iterations over every query predictions set. """
+
+    if indexes is None:
+        indexes = np.full_like(preds, fill_value=0, dtype=np.int64)
+    if isinstance(indexes, Tensor):
+        indexes = indexes.cpu().numpy()
+    if isinstance(preds, Tensor):
+        preds = preds.cpu().numpy()
+    if isinstance(target, Tensor):
+        target = target.cpu().numpy()
+
+    assert isinstance(indexes, np.ndarray)
+    assert isinstance(preds, np.ndarray)
+    assert isinstance(target, np.ndarray)
+
+    indexes = indexes.flatten()
+    preds = preds.flatten()
+    target = target.flatten()
+    groups = get_group_indexes(indexes)
+
     sk_results = []
+    for group in groups:
+        trg, pds = target[group], preds[group]
 
-    for b, a in zip(target, preds):
-        if b.sum() == 0:
-            if behaviour == 'skip':
+        if trg.sum() == 0:
+            if empty_target_action == 'skip':
                 pass
-            elif behaviour == 'pos':
+            elif empty_target_action == 'pos':
                 sk_results.append(1.0)
             else:
                 sk_results.append(0.0)
         else:
-            res = metric(b, a, **kwargs)
+            res = metric(trg, pds, **kwargs)
             sk_results.append(res)
 
     if len(sk_results) > 0:
         return np.mean(sk_results)
     return np.array(0.0)
 
 
-def _test_retrieval_against_sklearn(
-    sklearn_metric: Callable, torch_metric: Metric, size: int, n_documents: int, empty_target_action: str, **kwargs
-) -> None:
-    """ Compare PL metrics to standard version. """
-    metric = torch_metric(empty_target_action=empty_target_action, **kwargs)
-    shape = (size, )
+def _concat_tests(*tests: Tuple[str, Tuple]) -> Tuple[str, Tuple]:
+    """Concat tests composed by a string and a list of arguments."""
+    assert len(tests), "cannot concatenate "
+    assert all([tests[0][0] == x[0] for x in tests[1:]]), "the header must be the same for all tests"
+    return (tests[0][0], sum([x[1] for x in tests], []))
+
+
+_errors_test_functional_metric_parameters_default = [
+    "preds, target, message, metric_args",
+    [
+        # check input shapes are consistent (func)
+        (_irs_mis_sz_fn.preds, _irs_mis_sz_fn.target, "`preds` and `target` must be of the same shape", {}),
+        # check input tensors are not empty
+        (_irs_empty.preds, _irs_empty.target, "`preds` and `target` must be non-empty and non-scalar tensors", {}),
+        # check on input dtypes
+        (_irs.preds.bool(), _irs.target, "`preds` must be a tensor of floats", {}),
+        (_irs.preds, _irs.target.float(), "`target` must be a tensor of booleans or integers", {}),
+        # check targets are between 0 and 1
+        (_irs_bad_tgt.preds, _irs_bad_tgt.target, "`target` must contain `binary` values", {}),
+    ]
+]
+
+_errors_test_functional_metric_parameters_k = [
```

Comment:
this is parametrization for test, right, rather define it as dict so it clear what these arguments are

Reply:
ok, fixed, pls check the update...